core/finality-grandpa/src/communication/mod.rs

```diff
@@ -202,7 +211,15 @@ impl<B, S, H> Network<B> for Arc<NetworkService<B, S, H>> where
 	}
 }
 
-/// A stream used by NetworkBridge in its implementation of Network.
+/// A stream used by NetworkBridge in its implementation of Network. Given a oneshot that eventually returns a channel
+/// which eventually returns messages, instead of:
+///
+/// 1. polling the oneshot until it returns a message channel
+///
+/// 2. polling the message channel for messages
+///
+/// `NetworkStream` combines the two steps into one, requiring a consumer to only poll `NetworkStream` to retrieve
+/// messages directly.
 pub struct NetworkStream {
 	inner: Option<mpsc::UnboundedReceiver<network_gossip::TopicNotification>>,
 	outer: oneshot::Receiver<mpsc::UnboundedReceiver<network_gossip::TopicNotification>>
```

Comment:
note: this struct could be replace by an enum:
```rust
enum NetworkStream {
    PollingOneShot(oneshot::Receiver<mpsc::...>),
    PollingMpsc(mpsc::...),
}
```
showing the precise states of this stream.
But this is fine right now.